torch/csrc/jit/tensorexpr/schedule.cpp

```diff
@@ -399,6 +399,132 @@ class FunctionInliner : public IRMutator {
+// generated variable, and then bind those variables to rand() calls in the
+// body of the innermost for loop.
+class RandomInliner : public FunctionInliner {
+ public:
+  explicit RandomInliner(const std::vector<Function*>& funcs)
+      : FunctionInliner(funcs) {}
+
+  // Find the innermost for loop and bind any random variables in its body.
+  Stmt* mutate(const For* v) override {
+    const Var* var = v->var();
+    const Expr* start = v->start();
+    const Expr* stop = v->stop();
+    Stmt* body = v->body();
+    LoopOptions loop_options = v->loop_options();
+
+    // Bind the random variables in the innermost loop.
+    if (!is_inner_loop(v)) {
+      Stmt* new_body = body->accept_mutator(this);
+      if (new_body == body) {
+        return const_cast<For*>(v);
+      }
+      if (new_body == nullptr) {
+        return nullptr;
+      }
+      return new For(var, start, stop, new_body, loop_options);
+    }
+
+    // "Inline" all the random intrinsics by binding them up front and then
+    // using the bound variable in the body.
+    Stmt* new_body = Stmt::clone(body);
+    new_body = new_body->accept_mutator(this);
+    for (auto const& p : random_vars_) {
+      Var* v = p.second;
+      new_body = new LetStmt(v, new Intrinsics(kRand, v->dtype()), new_body);
+    }
+    if (new_body == body) {
+      return const_cast<For*>(v);
+    }
+    if (new_body == nullptr) {
+      return nullptr;
+    }
+    return new For(var, start, stop, new_body, loop_options);
+  }
+
+  // Inline calls containing rand().  Create a new random variable for each
+  // call being inlined, and remember which function is currently being inlined
+  // so we can look up the right variable to replace it with.
+  const Expr* mutate(const FunctionCall* v) override {
+    Function* prev_func = current_func_;
+    current_func_ = v->tensor()->function();
+
+    // Remember the calling args; if we find another call with different args,
+    // bail out because this case is too complicated.
+    auto it = call_args_.find(current_func_);
+    if (it == call_args_.end()) {
+      call_args_.emplace(current_func_, std::cref(v->params()));
+    } else {
+      if (v->params() != it->second.get()) {
+        throw std::runtime_error("Complex indexing pattern in rand() tensor");
+      }
+    }
+
+    // Assign a new random variable for this function, if needed.
+    if (!random_vars_.count(current_func_)) {
+      const std::string& name = current_func_->func_var(0)->name_hint();
+      random_vars_.emplace(current_func_, new Var(name, v->dtype()));
+    }
+    const Expr* result = FunctionInliner::mutate(v);
+    current_func_ = prev_func;
+    return result;
+  }
+
+  // Replace rand() intrinsics.
+  const Expr* mutate(const Intrinsics* v) override {
+    if (v->op_type() != kRand) {
+      return v;
+    }
+    if (!current_func_) {
+      return v;
+    }
+    auto it = random_vars_.find(current_func_);
+    if (it == random_vars_.end()) {
+      return v;
```

Comment:
Should this ever happen? Can we replace it with an assert checking that all `rand` calls are replaced with a var+let_stmt?

Reply:
Yeah, I think this can actually happen if you choose not to inline some functions containing rand() (not that it happens with our current fuser, but still)